File: bench/py/cpu_hash.py
# ============================================================================
# bench/py/cpu_hash.py — CPU-yoğun FNV-1a benchmark (Python)
#
# saQut referansı: bench/sqt/cpu_hash.sqt ile birebir aynı algoritma.
# Sözleşme: bench/README.md §"Cross-language determinizm".
#   - int = 32-bit wrap taklidi: her aritmetik adım sonrası & 0xFFFFFFFF
#   - LCG: x = x*1103515245 + 12345, seed 123456789
#   - FNV-1a: offset basis 2166136261 (^ r), prime 16777619
#   - CHECKSUM çıktısı signed int32 karşılığıdır (saQut çıktısıyla eşleşir)
# ============================================================================

import time

SIZE = 2048
ROUNDS = 3000

MASK = 0xFFFFFFFF
FNV_OFFSET = (-2128831035) & MASK  # = 2166136261
FNV_PRIME = 16777619


def to_signed32(v):
    v &= MASK
    return v - 0x100000000 if v >= 0x80000000 else v


def main():
    buf = []
    x = 123456789
    for _ in range(SIZE):
        x = (x * 1103515245 + 12345) & MASK
        buf.append(x)

    t0 = time.perf_counter()
    acc = 0
    for r in range(ROUNDS):
        h = FNV_OFFSET ^ r
        for b in buf:
            h = ((h ^ b) * FNV_PRIME) & MASK
        acc ^= h
    elapsed_ms = round((time.perf_counter() - t0) * 1000)

    print(f"CHECKSUM={to_signed32(acc)}")
    print(f"ELAPSED_MS={elapsed_ms}")


if __name__ == "__main__":
    main()

File: bench/php/cpu_hash.php
<?php
// ============================================================================
// bench/php/cpu_hash.php — CPU-yoğun FNV-1a benchmark (PHP)
//
// saQut referansı: bench/sqt/cpu_hash.sqt ile birebir aynı algoritma.
// Sözleşme: bench/README.md §"Cross-language determinizm".
//   PHP int64'tür; 32-bit taşan çarpma float'a düşer. Bu yüzden tüm
//   çarpmalar mul32() üzerinden (16-bit split multiplication) yapılır ve
//   her adım & 0xFFFFFFFF ile maskelenir. Çıktıda signed int32'ye çevrilir.
//   - LCG: x = mul32(x, 1103515245) + 12345 (masked), seed 123456789
//   - FNV-1a: offset basis 2166136261 (^ r), prime 16777619
// ============================================================================

const BENCH_SIZE = 2048;
const BENCH_ROUNDS = 3000;
const M32 = 0xFFFFFFFF;
const FNV_OFFSET_U = 2166136261;
const FNV_PRIME = 16777619;

/** 32-bit wrap çarpma: a*b mod 2^32 (16-bit parçalara bölerek). */
function mul32(int $a, int $b): int {
    $a &= M32;
    $b &= M32;
    $hi = (($a >> 16) * $b) & M32;      // üst 16 bit katkısı
    $lo = (($a & 0xFFFF) * $b) & M32;   // alt 16 bit katkısı
    return ((($hi << 16) + $lo) & M32);
}

function toSigned32(int $v): int {
    $v &= M32;
    return $v >= 0x80000000 ? $v - 0x100000000 : $v;
}

$buf = [];
$x = 123456789;
for ($i = 0; $i < BENCH_SIZE; $i++) {
    $x = mul32($x, 1103515245);
    $x = ($x + 12345) & M32;
    $buf[] = $x;
}

$t0 = microtime(true);
$acc = 0;
for ($r = 0; $r < BENCH_ROUNDS; $r++) {
    $h = FNV_OFFSET_U ^ $r;
    for ($i = 0; $i < BENCH_SIZE; $i++) {
        $h = mul32($h ^ $buf[$i], FNV_PRIME);
    }
    $acc ^= $h;
}
$elapsedMs = (int)round((microtime(true) - $t0) * 1000);

echo "CHECKSUM=" . toSigned32($acc) . "\n";
echo "ELAPSED_MS={$elapsedMs}\n";
